crates/wasi-provider/src/states/starting.rs

```diff
@@ -0,0 +1,139 @@
+use std::ops::Deref;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use log::{debug, info};
+use tokio::sync::Mutex;
+
+use kubelet::container::{Container, ContainerKey};
+use kubelet::pod::{key_from_pod, Handle};
+use kubelet::provider;
+use kubelet::state::prelude::*;
+use kubelet::volume::Ref;
+
+use crate::wasi_runtime::{self, HandleFactory, Runtime, WasiRuntime};
+use crate::PodState;
+
+use super::running::Running;
+
+fn volume_path_map(
+    container: &Container,
+    volumes: &HashMap<String, Ref>,
+) -> anyhow::Result<HashMap<PathBuf, Option<PathBuf>>> {
+    if let Some(volume_mounts) = container.volume_mounts().as_ref() {
+        volume_mounts
+            .iter()
+            .map(|vm| -> anyhow::Result<(PathBuf, Option<PathBuf>)> {
+                // Check the volume exists first
+                let vol = volumes.get(&vm.name).ok_or_else(|| {
+                    anyhow::anyhow!(
+                        "no volume with the name of {} found for container {}",
+                        vm.name,
+                        container.name()
+                    )
+                })?;
+                let mut guest_path = PathBuf::from(&vm.mount_path);
+                if let Some(sub_path) = &vm.sub_path {
+                    guest_path.push(sub_path);
+                }
+                // We can safely assume that this should be valid UTF-8 because it would have
+                // been validated by the k8s API
+                Ok((vol.deref().clone(), Some(guest_path)))
+            })
+            .collect::<anyhow::Result<HashMap<PathBuf, Option<PathBuf>>>>()
+    } else {
+        Ok(HashMap::default())
+    }
+}
+
+pub(crate) async fn start_container(
+    pod_state: &mut PodState,
+    pod: &Pod,
+    container: &Container,
+) -> anyhow::Result<kubelet::container::Handle<wasi_runtime::Runtime, wasi_runtime::HandleFactory>>
+{
+    let module_data = pod_state
+        .run_context
+        .modules
+        .remove(container.name())
+        .expect("FATAL ERROR: module map not properly populated");
+    let client = kube::Client::new(pod_state.shared.kubeconfig.clone());
+    let env = provider::env_vars(&container, pod, &client).await;
+    let args = container.args().clone().unwrap_or_default();
+    let container_volumes = volume_path_map(container, &pod_state.run_context.volumes)?;
+
+    let runtime = WasiRuntime::new(
+        container.name().to_owned(),
+        module_data,
+        env,
+        args,
+        container_volumes,
+        pod_state.shared.log_path.clone(),
+        pod_state.run_context.status_sender.clone(),
+    )
+    .await?;
+
+    debug!("Starting container {} on thread", container.name());
+    runtime.start().await
+}
+
+pub(crate) type ContainerHandleMap =
+    HashMap<ContainerKey, kubelet::container::Handle<Runtime, HandleFactory>>;
+
+#[derive(Default, Debug)]
+/// The Kubelet is starting the Pod containers
+pub(crate) struct Starting {
+    init_handles: Arc<Mutex<ContainerHandleMap>>,
```

Comment:
Note for other reviewers, this was necessary to have interior mutability as the `next` method is not `mut self`

Reply:
I'm curious why a state would need a mutable member in the design we have - isn't all the work and interior state essentially local to the `next` method?  How is this used that it could not be local to `next`?  (I'm not challenging you - I am trying to clarify my understanding of the state machine design.)

Reply:
We need to pass the current handles from the init container down the line. This is purely internal to this state. In this specific instance, because we need to take ownership of these handles, we must make the vec mutable so as to be able to drain it and take ownership of its contents. That clarify it at all?

Reply:
Can container handles be a part of `PodState`?

Reply:
It feels like an anti-pattern to have _anything_ that needs to be sent to a single specific state shoved into a shared pod state across the whole thing. We only need to forward these handles to the `Starting` state and do not need to track them otherwise. We did try to change the trait for `next` to be `self` instead of `&self` as it kinda makes sense for it to consume itself (as it always returns another state), but that gave the compiler problems. We could also trying making it `&mut self` though, which I can't believe I didn't think of doing.